File: dialingtest-service/src/main/java/com/huawei/cloududn/dialingtest/service/UserRoleService.java
package com.huawei.cloududn.dialingtest.service;

import com.huawei.cloududn.dialingtest.dao.UserRoleDao;
import com.huawei.cloududn.dialingtest.model.*;
import com.huawei.cloududn.dialingtest.util.OperationLogUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

/**
 * 用户角色管理服务
 */
@Service
@Transactional
public class UserRoleService {
    
    @Autowired
    private UserRoleDao userRoleDao;
    
    @Autowired
    private OperationLogUtil operationLogUtil;
    
    /**
     * 创建用户角色关系
     */
    public UserRole createUserRole(String username, String role, String operatorUsername) {
        // Validate role
        if (!isValidRole(role)) {
            throw new IllegalArgumentException("Invalid role: " + role);
        }
        
        // Check if already exists
        if (userRoleDao.existsByUsernameAndRole(username, role)) {
            throw new IllegalArgumentException("User role relationship already exists");
        }
        
        // Create user role relationship
        UserRole userRole = new UserRole();
        userRole.setUsername(username);
        userRole.setRole(UserRole.RoleEnum.fromValue(role));
        
        userRoleDao.insert(userRole);
        
        // Log operation to operation log module
        operationLogUtil.logUserRoleCreate(operatorUsername, userRole);
        
        return userRole;
    }
    
    /**
     * 分页查询用户角色关系
     */
    public UserRolePageResponseData getUserRolesWithPagination(int page, int size, String search) {
        int offset = page * size;
        List<UserRole> content = userRoleDao.findByUsernameContainingIgnoreCase(search, offset, size);
        int totalElements = userRoleDao.countByUsernameContainingIgnoreCase(search);
        int totalPages = (int) Math.ceil((double) totalElements / size);
        
        UserRolePageResponseData data = new UserRolePageResponseData();
        data.setContent(content);
        data.setTotalElements(totalElements);
        data.setTotalPages(totalPages);
        data.setSize(size);
        data.setNumber(page);
        data.setFirst(page == 0);
        data.setLast(page >= totalPages - 1);
        
        return data;
    }
    
    /**
     * 根据用户名获取用户角色
     */
    public List<String> getUserRolesByUsername(String username) {
        List<UserRole> userRoles = userRoleDao.findByUsername(username);
        return userRoles.stream()
                .map(userRole -> userRole.getRole().toString())
                .collect(Collectors.toList());
    }
    
    /**
     * 检查用户是否拥有指定角色
     */
    public boolean hasRole(String username, String role) {
        return userRoleDao.existsByUsernameAndRole(username, role);
    }
    
    /**
     * 检查用户是否拥有指定角色中的任意一个
     */
    public boolean hasAnyRole(String username, List<String> roles) {
        return userRoleDao.existsByUsernameAndRoleIn(username, roles);
    }
    
    /**
     * 更新用户角色关系
     */
    public UserRole updateUserRole(Integer id, String username, String role, String operatorUsername) {
        // Validate role
        if (!isValidRole(role)) {
            throw new IllegalArgumentException("Invalid role: " + role);
        }
        
        // Find existing record
        UserRole existingUserRole = userRoleDao.findById(id);
        if (existingUserRole == null) {
            throw new IllegalArgumentException("User role relationship does not exist");
        }
        
        String oldUsername = existingUserRole.getUsername();
        String oldRole = existingUserRole.getRole().toString();
        
        // If username or role changed, check if new username and role combination already exists (excluding current record)
        boolean usernameChanged = !oldUsername.equals(username);
        boolean roleChanged = !oldRole.equals(role);
        if (usernameChanged || roleChanged) {
            UserRole existingWithSameKey = userRoleDao.findByUsernameAndRole(username, role);
            if (existingWithSameKey != null && !existingWithSameKey.getId().equals(id)) {
                throw new IllegalArgumentException("User role already exists");
            }
        }
        
        // Update user role relationship
        existingUserRole.setUsername(username);
        existingUserRole.setRole(UserRole.RoleEnum.fromValue(role));
        
        userRoleDao.update(existingUserRole);
        
        UserRole newUserRole = new UserRole();
        newUserRole.setId(existingUserRole.getId());
        newUserRole.setUsername(username);
        newUserRole.setRole(UserRole.RoleEnum.fromValue(role));
        
        // Log operation
        operationLogUtil.logUserRoleUpdate(operatorUsername, existingUserRole, newUserRole);
        
        return existingUserRole;
    }
    
    /**
     * 删除用户角色关系
     */
    public void deleteUserRole(Integer id, String operatorUsername) {
        // Find existing record
        UserRole existingUserRole = userRoleDao.findById(id);
        if (existingUserRole == null) {
            throw new IllegalArgumentException("User role relationship does not exist");
        }
        
        String username = existingUserRole.getUsername();
        String role = existingUserRole.getRole().toString();
        
        // Delete user role relationship
        userRoleDao.deleteById(id);
        
        // Log operation
        operationLogUtil.logUserRoleDelete(operatorUsername, existingUserRole);
    }
    
    /**
     * 统计EXECUTOR角色的数量
     */
    public int getExecutorCount() {
        return userRoleDao.countByRole("EXECUTOR");
    }
    
    /**
     * 验证角色是否有效
     */
    private boolean isValidRole(String role) {
        return Arrays.asList("ADMIN", "OPERATOR", "BROWSER", "EXECUTOR").contains(role);
    }
}

File: frontend/src/services/userRoleService.js
import { createApiRequestConfig } from '../utils/apiUtils.js';

const API_BASE_URL = '/dialingtest/api';

/**
 * 用户角色服务类
 */
export class UserRoleService {
  
  /**
   * 获取用户角色列表（分页）
   * @param page 页码（从0开始）
   * @param size 每页大小
   * @param search 搜索关键词（可选）
   * @returns 分页的用户角色列表
   */
  static async getUserRolesWithPagination(page = 0, size = 10, search) {
    const params = new URLSearchParams({
      page: page.toString(),
      size: size.toString(),
    });
    
    if (search && search.trim()) {
      params.append('search', search.trim());
    }
    
    const response = await fetch(`${API_BASE_URL}/user-roles?${params}`, createApiRequestConfig('GET', undefined, false));
    
    if (!response.ok) {
      throw new Error(`获取用户角色列表失败: ${response.statusText}`);
    }
    
    const result = await response.json();
    if (!result.success) {
      throw new Error(result.message || 'Failed to get user roles');
    }
    
    return result.data;
  }
  
  /**
   * 创建用户角色
   * @param userRoleData 用户角色数据
   * @returns 创建的用户角色
   */
  static async createUserRole(userRoleData) {
    const response = await fetch(
      `${API_BASE_URL}/user-roles`,
      createApiRequestConfig('POST', userRoleData, true)
    );
    
    if (!response.ok) {
      const errorResult = await response.json();
      throw new Error(errorResult.message || `创建用户角色失败: ${response.statusText}`);
    }
    
    const result = await response.json();
    if (!result.success) {
      throw new Error(result.message || 'Failed to create user role');
    }
    
    return result.data;
  }
  
  /**
   * 更新用户角色
   * @param id 用户角色ID
   * @param userRoleData 用户角色数据
   * @returns 更新后的用户角色
   */
  static async updateUserRole(id, userRoleData) {
    const response = await fetch(
      `${API_BASE_URL}/user-roles/${id}`,
      createApiRequestConfig('PUT', userRoleData, true)
    );
    
    if (!response.ok) {
      const errorResult = await response.json();
      throw new Error(errorResult.message || `更新用户角色失败: ${response.statusText}`);
    }
    
    const result = await response.json();
    if (!result.success) {
      throw new Error(result.message || 'Failed to update user role');
    }
    
    return result.data;
  }
  
  /**
   * 删除用户角色
   * @param id 用户角色ID
   */
  static async deleteUserRole(id) {
    const response = await fetch(
      `${API_BASE_URL}/user-roles/${id}`,
      createApiRequestConfig('DELETE', undefined, true)
    );
    
    if (!response.ok) {
      const errorResult = await response.json();
      throw new Error(errorResult.message || `删除用户角色失败: ${response.statusText}`);
    }
  }
  
  
  
  /**
   * 获取EXECUTOR角色数量
   * @returns EXECUTOR角色数量
   */
  static async getExecutorCount() {
    const response = await fetch(
      `${API_BASE_URL}/user-roles/executor-count`,
      createApiRequestConfig('GET', undefined, false)
    );
    
    if (!response.ok) {
      throw new Error(`获取执行机数量失败: ${response.statusText}`);
    }
    
    const result = await response.json();
    if (!result.success) {
      throw new Error(result.message || 'Failed to get executor count');
    }
    
    return result.data;
  }
}
